README
Kai Zhang  <kay21s AT gmail DOT com> /* old version Author */
Yanfei Wang <backyes AT gmail DOT com>

Based on IOEngine from PacketShader, send packets from a specified trace file at speed of 10Gbps
Make some modification in the driver to adapt to current kernel version(2.6.37)

Improvements:
/* Author: Yanfei Wang, <backyes@gmail.com>
 * Fri Mar 16 18:34:00 CST 2012
 * improvements
 * (1) send packets at the same time for all queueus.
 * (2) never discard any packets for specified traces
 * (3) allow loop or once-trace-replay as you wish.
 * (4) use pthread lib to rewrite it for supporting fine grained sysnc.
 * optimization on performance
 * (5) fix bugs on spin_lock_bh spothot(>80% overhead on cpu cycle).
 * (6) supports on numa lib, allocate memory after numa bind.
 * (7) do sync. among all threads with pthread lib instead of "fork", so rewrite most data structure for per-thread variable.
 */
Improments:
/* Author: Yanfei Wang, <backyes@gmail.com>
 *	(1) usleep(n) to avoid small chunk size cause by sending remained packets for keeping "non-discard replay " manner..
 * (2) performance test 
 *	fix.mill-->split0.pcap split1.pcap ...(total 8 sub-traces)
 *	average packet size: ~90B (non-64Byte)
 *	Performance: 7.9Gbps(L2 payload), 10.4Mpps...
 * (3) Conclusion:
 *	it can replay trace at the speed of approximately line-rate...
 */
add-on: Osplit      --one program that split trace file based on TCP flow.
	Since the speed gap between the disk and memory, we must buffer the disk trace file in kenrel
for high speed. Also, to take advantage of multi-queue for 10G replay, we expect to split the raw
trace into N child traces based on ways, such as src IP + dst IP, 4 tuples srcIP+dstIP+srcPort+dstPort, to keep
flow affinity and TX queue balance at best. 
The Split program is in directory of userspace_split,
	Usage of Oplist:
	1. install the libnet-1.1.5 in sub-directroy of userspace_split
	2. install the libpcap-1.1.1 in sub-directroy of userspace_split
	3. install the libnids-1.24 in sub-directroy of userspace_split
	modify line 73 in Osplit.c which locates in src directory to locate the child traces.
	cd libnids-1.24
	./configure
	make
	cd src

NOTICE:

driver/Makefile  	-- Set "KSRC" to the kernel dir
interface name	 	-- IOEngine uses ixg%d, while our machine uses eth%d, can be modified in driver/install.py
sample/tx/tx.c    -- BATCH_SIZE, adjust to reduce system call overhead

USAGE:

Each TX thread uses a trace file, specify them in sample/tx/pktbuf.c : char fname[QUEUE_NUM][256];

./configure
Make
./install.py [# of RX queue] [# of TX queue]
./sample/tx/rep_trace [loop] xge# /* loop indicate replay trace in loop way. */




numactl-2.0.8-rc4/CHANGES
[Note: newer entries are at the bottom]

The actual patches and their series are in the tarball:  patches.tar.gz

Early prehistory:

- Fix some warnings
- Add const to global variables and prototypes
- Document numa_warn
- Add numa_distance() support to read topology
- remove internal alias from numa_{warn,error} to allow overwriting
  again
- Various bug fixes
- add -f option to memhog to map file
- more bugfixes
- replace forward backward with STREAM test in numademo

0.6.1

- make headers C++ clean

0.6.2 

- use more accurate buffer length for cpumask
- add --cpubind to test suite

0.6.3

- fix cpumask parser for large number of cpus.
Note that you need a kernel patch (as of 2.6.7-pre) for that too
if the cpumask is longer than 99 characters.

0.6.4

- Add Coppyright headers

0.6.5

- Reduce unneeded DSO relocations (Arjan van de Ven) 
- Add -r option to memhog (repeat walk) 
- some manpage fixes
- Use syscall numbers from asm/unistd.h if possible
- Add numa_node_size64 to handle large nodes on 32bit architectures
- Fix numactl to use it (report from Rajan Ravindran) 
- Use ln -sf in make install (Rajan Ravindran)
- Add syscall numbers for ppc/ppc64
- Add private syscall6 for i386 since the glibc version is broken
- Remove STUB
- Change numactl --show to use cpubind instead of nodebind for CPU affinity.
- Fix make install into examples directory
- Work around broken sched_set_affinity API. This adds a 32768 CPUs limit.
- Fix segfault with /sys not mounted.
- Clean up Makefile
- Make numactl --show more clever

0.7-pre1

- add test/regress2 and some fixes to test programs
- Fix DSO relocation patch for global variables
- Change nodeset sizes back to be binary compatible with SLES9
- Cosmetic changes to manpages (pointed out by Eric S. Raymond) 
- Make numa_run_on_node etc. act on current thread only even on NPTL systems
  (Dinakar Guniguntala)
- Make numa_no_nodes / numa_all_nodes const (Werner Almesberger) 
- Fix up the warnings caused by above change
- Add numa_distance() on systems with ACPI
- remove some obsolete code
- add rdtsc for ppc64
- fix unsigned/unsigned long confusion in cpumasks (Matt Dobson)
- fix CPU_BYTES and rename CPU_WORDS to CPU_LONGS (Matt Dobson) 
- Print node distances in numactl

[0.7 skipped]

0.8
- hardend numactl command line parsing against bad arguments in some cases
- remove cpumask/nodemask confusion which has become a FAQ:
  --cpubind deprecated, added --cpunodebind and --physcpubind= options
  print both in --show, old cpumask kept for compatibility
- Fix --show problems
- various fixes for bugs noted by Mike Stroyan (thanks!)
- install set_mempolicy manpage
- various smaller fixes

0.9
- Get rid of bogus distance.o that broke compilation on !x86-64 (sorry)
- Handle CFLAGS overriding without OPT_CFLAGS (Ian Wienand)
- Fix up section of get/set_mempolicy (Ian Wienand)
- When no NUMA available fall back to one global node instead of one node
per CPU (Samuel Thibault)
- Don't rely on architecture symbols for dependency generation
- Use __powerpc__ to detect PPC/PPC64
- numastat: 
  * wrap display properly with many nodes
  * display nodes in forward order
  * install manpage in `make install'.
- remove bogus numamemcpy.c
- numademo: 
  * allow standalone compile, make streamlib optional
  * clean up output
  * change output unit to standard MB/s
  * compile with more optimization
  * add random pass to fool any prefetching (slow)
- make numademo compileable outside source tree
- use gettimeofday instead of time stamp counters in benchmarks
- support valgrind in testsuite
- other minor changes

0.9.1
- Make automatic selection of lib64 vs lib more robust. Now should work
even on ppc32 with a lib64 directory. Architecture lists are hardcoded now
unfortunately.

0.9.2
- Fix compilation on architectures with gcc 3.3+ but without TLS
(MIPS, Alpha, Sparc) 
- Add warning against using of MPOL_F_NODE
- numa.3 improvements from Michael Kerrisk
- Support page migration (migratepages, manpages) from Christoph
Lameter. Requires 2.6.16+ kernels

0.9.3
- Some more manpage fixes
- install migratepages manpage in make install
- Build fix for Debian make from Ian Wienand

0.9.4:
- Remove syscall manpages. They're in main man-pages now.
- More migrate fixes from C.Lameter

0.9.5:
- Fix parsing of cpumap in sysfs from Doug Chapman

0.9.6:
- Fix make install again

0.9.7
- Fix cpumap parsing fix to not corrupt memory (Doug Chapman) 
- Small optimization to cpumap parsing
- Create target directories for Debian (Ian Wienand)

0.9.8
- Fix cpumap parsing again (Doug Chapman)

0.9.9 (aka "Will 1.0 ever happen?") 
- Fix sizing of cpu buffers for affinity syscalls
- Don't corrupt errno in numa_run_on_node_mask. This fixes
numactl cpubind issues on some systems.
- Print cpus belonging to nodes in numactl --hardware
- Rewrite cpumap parser to be simpler and hopefully finally work
in all cases
- add testcases for cpu affinity and topology discovery
- Add make test target to run regression test easier and fix up
test/README

Lots of fixes thanks to thorough testing by Noriyuki Taniuchi:
- Better command line parsing in numactl and fix various documentation bugs
- Wrong arguments to --prefered don't crash numactl anymore
- Fix --cpunodebind=all
- Auto collect short option list in numactl - a couple were missing.
- Fix documentation of numa_set_localalloc. It doesn't have a flag. 
- Fix numa_run_on_node(-1)
- numa_get_run_node_mask(): Fix documentation, don't warn

0.9.10:
- Fix cpumap parsing bug when NR_CPUS < 32 (dean gaudet)  

0.9.11
- Fix usage output for --shmid (Noriyuki Taniuchi)
- Use correct syscall number for migrate_pages() on PPC
 
1.0
- Add sleep to regression test to work with delayed statistic
updating in newer kernels (Mel Gorman) 
- Default to -O2

1.0.1
- Fix build on powerpc

1.0.2
- Fix parallel Makefile build (Andreas Herrmann)
- Fix target command argument parsing for numactl (no -- needed again anymore) 
- Clarify numa_node_to_cpus() manpage

1.0.3


samples/tx/tx.c
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#define __USE_GNU

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sched.h>
#include <unistd.h>
#include <assert.h>
#include <errno.h>
#include <string.h>

#include <sys/wait.h>
#include <sys/time.h>
#include "pkt_buff.h"
#include <pthread.h>
#include <numa.h>

#include "../../include/ps.h"

/* Author: Yanfei Wang, backyes@gmai.com
 * Fri Mar 16 18:34:00 CST 2012
 * improvements
 * (1) send packets at the same time for all queueus.
 * (2) never discard any packets for specified traces
 * (3) allow loop or once-trace-replay as you wish.
 * (4) use pthread lib to rewrite it for supporting fine grained sysnc.
 * optimization on performance
 * (5) fix bugs on spin_lock_bh spothot that caused by wrong program.
 * (6) supports on numa lib, allocate memory after numa bind.
 */
#define MAX_CPUS 32
/* chunk batch size */
#define BATCH_SIZE 128

/* global var */
int num_devices;
struct ps_device devices[MAX_DEVICES];
int num_devices_attached;
int devices_attached[MAX_DEVICES];
double tot_gbps[MAX_CPUS];
long tot_packets[MAX_CPUS];
double tot_mpps[MAX_CPUS];
int loop = 0;

/* thread worker var */
pthread_barrier_t  barrier;

typedef struct _thread_ctrl {
	pthread_t th_ctrl;
	pthread_barrier_t *barrier;
	struct timeval starttime;
	struct timeval endtime;
	int cpu_id;
	int num_devices_attached;
	int devices_attached[MAX_DEVICES];
	struct ps_handle handle;

}thread_ctrl;

thread_ctrl th_ctrl[MAX_CPUS];
pthread_barrier_t barrier;

int get_num_cpus()
{
	return sysconf(_SC_NPROCESSORS_ONLN);
}

int bind_cpu(int cpu)
{
   cpu_set_t *cmask;
	struct bitmask *bmask;
	size_t n;
	int ret;

	n = get_num_cpus();

	if (cpu < 0 || cpu >= (int)n) {
		errno = -EINVAL;
		return -1;
	}
	cmask = CPU_ALLOC(n);
	if (cmask == NULL)
		return -1;

	CPU_ZERO_S(n, cmask);
	CPU_SET_S(cpu, n, cmask);

	ret = sched_setaffinity(0, n, cmask);
	usleep(1000);
	/* skip NUMA stuff for UMA systems */
	if (numa_max_node() == 0)
		return ret;

	bmask = numa_bitmask_alloc(16);
	assert(bmask);

	numa_bitmask_setbit(bmask, cpu % 2);
	numa_set_membind(bmask);
	numa_bitmask_free(bmask);

	return ret;
}

void print_usage(char *argv0)
{
	fprintf(stderr, "Usage: %s <interface to echo> <...>",
			argv0);

	exit(2);
}

void parse_opt(int argc, char **argv)
{
	int i, j;
	int next;

	if (argc < 2)
		print_usage(argv[0]);
	
	if(!strcmp(argv[1], "loop")) {
		loop = 1;
		next = 2;
	}
	else next = 1;

	for (i = next; i < argc; i++) {
		int ifindex = -1;

		for (j = 0; j < num_devices; j++) {
			if (strcmp(argv[i], devices[j].name) != 0)
				continue;

			ifindex = devices[j].ifindex;
			break;
		}

		if (ifindex == -1) {
			fprintf(stderr, "Interface %s does not exist!\n", argv[i]);
			exit(4);
		}

		for (j = 0; j < num_devices_attached; j++) {
			if (devices_attached[j] == ifindex)
				goto already_attached;
		}

		devices_attached[num_devices_attached] = ifindex;
		num_devices_attached++;

already_attached:
		;
	}

	assert(num_devices_attached > 0);
}

void *echo(void *arg)
{
	thread_ctrl *pth_ctrl = (thread_ctrl*) arg;
	int cpu_id = pth_ctrl->cpu_id;
	struct ps_handle *handle = &pth_ctrl->handle;
	int num_devices_attached = pth_ctrl->num_devices_attached;
	int* devices_attached = pth_ctrl->devices_attached;
	struct ps_chunk chunk;
	//!trace variable
	file_cache_t *fct;
	unsigned int pktlen;
	u_char *pktdata;
	//!performance variable
	uint64_t total_tx_packets = 0;
	uint64_t total_tx_bytes = 0;
	struct timeval subtime;
	int ifindex;

	int i;
	int working = 0;
	int loop_end = 0;

	bind_cpu(cpu_id);
	assert(ps_init_handle(handle) == 0);

	//Preload pcap file --must done after bind_cpu() call. 
	if ((fct = preload_pcap_file(cpu_id)) != NULL) {
		printf("Loading done, core %d\n", cpu_id);
		if (!check_pcap(fct))
			printf("It is not trace file, core %d\n", cpu_id);
	} else {
		printf("Loading failed, core %d\n", cpu_id);
	}

	for (i = 0; i < num_devices_attached; i++) {

		if (devices[devices_attached[i]].num_tx_queues <= cpu_id) {
			printf("WARNING: xge%d has not enough TX queues!\n",
					devices_attached[i]);
			continue;
		}

		working = 1;

		printf("attaching TX queue xge%d:%d to CPU%d\n", devices_attached[i], cpu_id, cpu_id);
	}

	if (!working)
		goto done;

	assert(ps_alloc_chunk(handle, &chunk) == 0);
	
	/* initialise chunk */
	chunk.queue.qidx = cpu_id;
	chunk.cnt = BATCH_SIZE;
	chunk.recv_blocking = 1;

	printf("%d CPU: chunk.cnt = %d\n", cpu_id, chunk.cnt);
	
	/* begin to replay */
	pthread_barrier_wait(pth_ctrl->barrier);

	gettimeofday(&(pth_ctrl->starttime), NULL);

	while(1) {

		chunk.cnt = BATCH_SIZE;

		/* build packets */
		for (i=0; i < chunk.cnt; i++) {
retry:
			pktdata = prep_next_skb(fct, &pktlen);

			if(pktdata == NULL) {
				if(loop == 0) { 
					loop_end = 1;
					goto last_chunk;
				}
				else {
					fct->offset = sizeof(pf_hdr_t);
					goto retry;
				}
			}

			chunk.info[i].offset = i * MAX_PACKET_SIZE;
			chunk.info[i].len = pktlen;
			memcpy_aligned_tx(chunk.buf + chunk.info[i].offset,
					pktdata,
					pktlen);
		}

last_chunk:
		/* set cnt */
		chunk.cnt = i;

		int ret;
		/* send packets */
		for (i = 0; i < num_devices_attached; i++) {

			chunk.queue.ifindex = devices_attached[i];

send_left:
			ret = ps_send_chunk(handle, &chunk);

			assert(ret >= 0);
			
			/* never discard one packet */
			if(ret < chunk.cnt) {
				for(i = ret; i < chunk.cnt; i++) {
					chunk.info[i - ret].offset = (i - ret) * MAX_PACKET_SIZE;
					chunk.info[i - ret].len = pktlen;
					memcpy_aligned_tx(chunk.buf + chunk.info[i - ret].offset,
							chunk.buf + chunk.info[i].offset,
							chunk.info[i].len);
				}
				chunk.cnt -= ret;
				assert(chunk.cnt >= 0);
				usleep(10);
				goto send_left;
			} /* end resend packets left */

		}
		/* send over */
		if(loop_end == 1) break;

		/* dynamic performance */
		if(loop == 1) {
			total_tx_packets = 0;
			total_tx_bytes = 0;
			for (i = 0; i < num_devices_attached; i++) {
				ifindex = devices_attached[i];
				total_tx_packets += handle->tx_packets[ifindex];
				total_tx_bytes += handle->tx_bytes[ifindex];
			}
			if(total_tx_packets % 2000000 < 100) { 

				gettimeofday(&(pth_ctrl->endtime), NULL);
				timersub(&pth_ctrl->endtime, &pth_ctrl->starttime, &subtime);

				printf("CPU %d: %ld packets transmitted, elapse time : %lds/%ldus Send Speed : %lf Mpps, %5.2f Gbps(payload), Aveage Len. = %ld\n", 
						cpu_id, total_tx_packets, subtime.tv_sec, subtime.tv_usec, 
						(double)(total_tx_packets) / (double) (subtime.tv_sec*1000000+subtime.tv_usec),
						(double)(total_tx_bytes*8) / (double) ((subtime.tv_sec*1000000+subtime.tv_usec) * 1000),
						total_tx_bytes/total_tx_packets);
			}
			
		} /* end performance */ 

	} /* external loop */
	
	gettimeofday(&(pth_ctrl->endtime), NULL);
	timersub(&pth_ctrl->endtime, &pth_ctrl->starttime, &subtime);

	printf("%d CPU: last chunk %d packets\n", cpu_id, chunk.cnt);

	/* performance statistics */
	total_tx_packets = 0;
	total_tx_bytes = 0;
	for (i = 0; i < num_devices_attached; i++) {
		ifindex = devices_attached[i];
		total_tx_packets += handle->tx_packets[ifindex];
		total_tx_bytes += handle->tx_bytes[ifindex];
	}

	printf("CPU %d: %ld packets transmitted, elapse time : %lds/%ldus Send Speed : %lf Mpps, %5.2f Gbps(payload), Aveage Len. = %ld\n", 
			cpu_id, total_tx_packets, subtime.tv_sec, subtime.tv_usec, 
			(double)(total_tx_packets) / (double) (subtime.tv_sec*1000000+subtime.tv_usec),
			(double)(total_tx_bytes*8) / (double) ((subtime.tv_sec*1000000+subtime.tv_usec) * 1000),
			total_tx_bytes/total_tx_packets);

	
	tot_packets[cpu_id] = total_tx_packets;
	tot_gbps[cpu_id] = (double)(total_tx_bytes*8) / (double) ((subtime.tv_sec*1000000+subtime.tv_usec) * 1000);
	tot_mpps[cpu_id] = (double)(total_tx_packets) / (double) (subtime.tv_sec*1000000+subtime.tv_usec);

	for (i = 0; i < num_devices_attached; i++) {
		char *dev = devices[devices_attached[i]].name;
		ifindex = devices_attached[i];

		if (handle->tx_packets[ifindex] == 0)
			continue;

		printf("  %s: ", dev);
		
		printf("TX %ld packets "
				"(%ld chunks, %.2f packets per chunk)\n", 
				handle->tx_packets[ifindex],
				handle->tx_chunks[ifindex],
				handle->tx_packets[ifindex] / 
				  (double)handle->tx_chunks[ifindex]);
	}

done:
	ps_close_handle(handle);
	return 0;
}


int main(int argc, char **argv)
{
	int num_cpus;
	int i ;

	num_cpus = get_num_cpus();
	assert(num_cpus >= 1);

	num_devices = ps_list_devices(devices);
	if (num_devices == -1) {
		perror("ps_list_devices");
		exit(1);
	}

	parse_opt(argc, argv);

	pthread_barrier_init(&barrier, NULL, num_cpus);

	for (i = 0; i < num_cpus; i ++) { 
		th_ctrl[i].cpu_id = i;
		th_ctrl[i].barrier = &barrier;
		th_ctrl[i].num_devices_attached = num_devices_attached;
		memcpy(th_ctrl[i].devices_attached, devices_attached, sizeof(int) * MAX_DEVICES);
		int ret = pthread_create(&(th_ctrl[i].th_ctrl), NULL, echo, (void*)&(th_ctrl[i]));
		assert(ret == 0);	
	}

	for(i = 0; i < num_cpus; i++)
		pthread_join(th_ctrl[i].th_ctrl, NULL);

	long packets = 0;
	double gbps = 0;
	double mpps = 0;
	for(i = 0; i < num_cpus; i++) { 
		packets += tot_packets[i];
		gbps += tot_gbps[i];
		mpps += tot_mpps[i];
	}
	printf("-----------------------\n");
	printf("tot_packets: %ld, tot_gbps: %lf, tot_mpps: %lf\n", packets, gbps, mpps);
	return 0;
}
